### DISCUSSION TITLE
Is there a way to defer instead of running state?

### DISCUSSION THREAD

**Discussion by tanaga9** [apache/airflow#31176] [Q&A]
Background (purpose)
-------------------------

Part of the workflow requires manual intervention under certain conditions.

Tried
---------

When a certain condition,
Wait for manual intervention using `TimeDeltaSensorAsync`.
Resume later.

Problem
---------

Tasks in the `deferred` state gradually accumulate,
It consumes the DAG's `max_active_runs` and lowers throughput.
Temporarily, nothing works, even when idle.
(all running tasks in `deferred` state)
(work !)

Question
-----------

Is there a way to defer instead of running state?

**Comment by potiuk** [comment]
Yes. You have to write a deferrable operator/Trigger for that or use some of an existing ones.

**Comment by tanaga9** [comment]
thanks @potiuk 

I used `TimeDeltaSensorAsync`.

- https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/time.html#timesensorasync
- https://github.com/apache/airflow/blob/c3c90eaa949b34c18f7b49052b2733ed817f8bed/airflow/sensors/time_delta.py#L50

`TimeDeltaSensorAsync` is a deferrable operator.
(it is not?)
I've seen a lot of tasks go into the `deferred` state.

Tasks in the `deferred` state count towards active runs (consume max_active_runs).
When tasks in the `deferred` state accumulate to max_active_runs, nothing will work.
Is it a feature and not a bug?

**Reply by potiuk** [reply]
I believe so. We've discussed it in the past (can't find answer) but @andrewgodwin can likely confim it that yes - deferred task still consume pools and other counters.

**Reply by andrewgodwin** [reply]
Deferrable tasks are only meant to count against pools/active count while they are active, not while they are deferred in a triggerer, if I recall correctly. It's been a while though!

**Reply by tanaga9** [reply]
@potiuk , @andrewgodwin 

A task in `deferred` state by `TimeDeltaSensorAsync` was handled as follows.

| Task State  | DAG's  running ? (counts active_runs ?) | consume pools ? |
| ----------- | ------------- | --------------- |
| `deferred`  | yes           | no              |
| `queued`    | no            | no              |
| `scheduled` | yes           | no              |
| `running`   | yes           | yes             |
| `success`   | no            | no              |
| `failed`    | no            | no              |

At least such a table should be in the document.

The following points are difficult to understand.

- What is `running` ?
- What is `active_runs` ? (Which state affects DAG's `max_active_runs` )
- Task state `running` is **not**  DAG's `running`. (different)
    - DAG's `running` contains  "Task state `deferred` and `scheduled` and `running`"

**Reply by tanaga9** [reply]
What is [`max_active_runs`](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/dag/index.html) ?
> max_active_runs (int) – maximum number of active DAG runs, beyond this number of DAG runs in a running state, the scheduler won’t create new active DAG runs

> DAG runs in a running state

When task status is either `deferred` or `scheduled` or `running`

**Comment by astronautas** [comment]
The issue still seems to be there :/
